jormungandr/src/watch_client.rs

```diff
@@ -0,0 +1,329 @@
+    }
+}
+
+type SubscriptionTryStream<S> =
+    MapErr<S, fn(<S as TryStream>::Error) -> chain_network::error::Error>;
+type SubscriptionStream<S> =
+    Map<S, fn(<S as Stream>::Item) -> Result<<S as Stream>::Item, chain_network::error::Error>>;
+
+#[tonic::async_trait]
+impl Watch for WatchClient {
+    type BlockSubscriptionStream = SubscriptionTryStream<BroadcastStream<Block>>;
+    type TipSubscriptionStream = SubscriptionStream<WatchStream<Header>>;
+    type SyncMultiverseStream = SubscriptionTryStream<ReplyStream<Block, intercom::Error>>;
+
+    #[instrument(skip(self))]
+    async fn block_subscription(
+        &self,
+    ) -> Result<Self::BlockSubscriptionStream, chain_network::error::Error> {
+        let block_receiver = BroadcastStream::new(self.block_sender.subscribe());
+
+        // there are two possible errors for the block_receiver.
+        // one occurs when there are no more senders, but that won't happen here.
+        // the other is when the receiver is lagging.
+        let live_stream: SubscriptionTryStream<BroadcastStream<Block>> =
+            block_receiver.map_err(|e| chain_network::error::Error::new(Code::Aborted, e));
+
+        Ok(live_stream)
+    }
+
+    #[instrument(skip(self))]
+    async fn tip_subscription(
+        &self,
+    ) -> Result<Self::TipSubscriptionStream, chain_network::error::Error> {
+        let tip_receiver: SubscriptionStream<_> = WatchStream::new(self.tip_receiver.clone())
+            .map::<Result<Header, chain_network::error::Error>, _>(Ok);
+
+        Ok(tip_receiver)
+    }
+
+    #[instrument(skip(self))]
+    async fn sync_multiverse(
+        &self,
+        from: BlockIds,
+    ) -> Result<Self::SyncMultiverseStream, chain_network::error::Error> {
+        let (handle, future) = intercom::stream_reply(32);
+
+        self.request_tx
+            .clone()
+            .send(RequestMsg::SyncMultiverse { from, handle })
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Unavailable, e))?;
+
+        let stream = future
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Internal, e))?;
+
+        Ok(stream.map_err(|e| chain_network::error::Error::new(Code::Internal, e)))
+    }
+}
+
+async fn handle_sync_multiverse(
+    checkpoints: BlockIds,
+    blockchain: &Blockchain,
+    storage: &Storage,
+    sink: &mut intercom::ReplyStreamSink<Block>,
+) -> Result<(), intercom::Error> {
+    let mut checkpoints = checkpoints
+        .iter()
+        .map(|id| {
+            let deser = HeaderHash::deserialize(id.as_bytes())
+                .map_err(intercom::Error::invalid_argument)?;
+
+            let chain_length = storage
+                .get_chain_length(deser)
+                .ok_or_else(|| intercom::Error::invalid_argument("unknown checkpoint"))?;
+
+            Ok::<_, intercom::Error>((chain_length, deser))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
```

Comment:
I think this will actually result in a an error if a checkpoint is unknown to the node, which is not what we want. Probably `filter_map` is what we should be using here